tests/beta/tapes/test_tape.py

```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
 
         expected = [None, [1, -1, -1, 1], [0, 5]]
         assert [m.eigvals is r for m, r in zip(new_tape.measurements, expected)]
+
+
+class TestExecution:
+    """Tests for tape execution"""
+
+    def test_execute_parameters(self, tol):
+        """Test execution works when parameters are both passed and not passed."""
+        dev = qml.device("default.qubit", wires=2)
+        params = [0.1, 0.2]
+
+        with QuantumTape() as tape:
+            qml.RX(params[0], wires=[0])
+            qml.RY(params[1], wires=[1])
+            qml.CNOT(wires=[0, 1])
+            expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        assert tape.output_dim == 1
+
+        # test execution with no parameters
+        res1 = tape.execute(dev)
+        assert tape.get_parameters() == params
+
+        # test execution with parameters
+        res2 = tape.execute(dev, params=[0.5, 0.6])
+        assert tape.get_parameters() == params
+
+        # test setting parameters
+        tape.set_parameters(params=[0.5, 0.6])
+        res3 = tape.execute(dev)
+        assert np.allclose(res2, res3, atol=tol, rtol=0)
+        assert not np.allclose(res1, res2, atol=tol, rtol=0)
+        assert tape.get_parameters() == [0.5, 0.6]
+
+    def test_no_output_execute(self):
+        """Test that tapes with no measurement process return
+        an empty list."""
+        dev = qml.device("default.qubit", wires=2)
+        params = [0.1, 0.2]
+
+        with QuantumTape() as tape:
+            qml.RX(params[0], wires=[0])
+            qml.RY(params[1], wires=[1])
+
+        res = tape.execute(dev)
+        assert res.size == 0
+        assert np.all(res == np.array([]))
+
+    def test_incorrect_output_dim_estimate(self):
+        """Test that a quantum tape with an incorrect output dimension
+        estimate corrects itself after evaluation."""
+        dev = qml.device("default.qubit", wires=3)
+        params = [1.0, 1.0, 1.0]
+
+        with QuantumTape() as tape:
+            qml.RX(params[0], wires=[0])
+            qml.RY(params[1], wires=[1])
+            qml.RZ(params[2], wires=[2])
+            qml.CNOT(wires=[0, 1])
+            probs(wires=0)
+            probs(wires=[1])
+
+        # estimate output dim should be correct
+        assert tape.output_dim == sum([2, 2])
+
+        # modify the output dim
+        tape._output_dim = 2
+
+        res = tape.execute(dev)
+        assert tape.output_dim == sum([2, 2])
+
+    def test_incorrect_ragged_output_dim_estimate(self):
+        """Test that a quantum tape with an incorrect *ragged* output dimension
```

Comment:
Not completely sure about the distinction between this test case and the previous one. Perhaps might be worth parametrizing?

Reply:
The difference is in the tape measurements 🙂  The previous test had output `[probs(wires=0), probs(wires=1)]` which will output a NumPy array with shape `(2,2)`.

This test instead has a tape with output `[probs(wires=0), probs(wires=[1, 2])]`, which will be a _ragged_ NumPy object array that looks like `[[a, b], [c, d, e, f]]`. So we are dealing with a completely different `dtype` to the previous case.

Since we are testing different logic here, I don't think this should be parametrized (parametrization should be restricted to testing the same logic with different data).